### PR TITLE
Format all patterns (well, except macros)

### LINKED ISSUE #18: Format patterns

### LINKED ISSUE #672: Rustfmt doesn't make the whitespace consistent in structs.
The following code is unchanged by rustfmt:

```rust
fn foo() {
    match instr {
        Set { amount, offset} => 1,
        PointerIncrement {amount, .. } => 2,
    }
}
```

Whereas I was expecting something like:

```rust
fn foo() {
    match instr {
        Set { amount, offset } => 1,
        PointerIncrement { amount, .. } => 2,
    }
}
```

### PR DESCRIPTION
**Pull request by nrc**
Fixes #18
Fixes #672 

r? @marcusklaas

### CODE REVIEW

File: src/patterns.rs [hunk 1]
```diff
@@ -55,39 +65,105 @@ impl Rewrite for Pat {
...
+                let slice_pat = slice_pat.as_ref().map(|p| {
+                    Some(format!("{}..", try_opt!(p.rewrite(context, width, offset))))
+                });
+                let suffix = suffix.iter().map(|p| p.rewrite(context, width, offset));
+
+                // Munge them together.
+                let pats = prefix.chain(slice_pat.into_iter()).chain(suffix);
+
+                // Check that all the rewrites succeeded, and if not return None.
+                let (somes, nones) = pats.partition::<Vec<Option<String>>, _>(Option::is_some);
```
Comment (marcusklaas) [inline src/patterns.rs]: I think we can `collect` into `Option<Vec<String>>` here.

File: src/patterns.rs [hunk 2]
```diff
@@ -55,39 +65,105 @@ impl Rewrite for Pat {
...
+                if nones.len() > 0 {
+                    return None;
+                }
+
+                // Unwrap all the sub-strings and join them with commas.
+                let pats = somes.into_iter().map(|p| p.unwrap()).collect::<Vec<_>>().join(", ");
+                Some(format!("[{}]", pats))
```
Comment (marcusklaas) [inline src/patterns.rs]: Are we sure we're staying within the bounds here? If not, we should call `wrap_str`.

File: src/patterns.rs [hunk 3]
```diff
@@ -55,39 +65,105 @@ impl Rewrite for Pat {
...
+                    (", ..", "..")
+                } else {
+                    ("", "}")
+                };
+
+                let budget = try_opt!(width.checked_sub(path.len() + 5 + elipses_str.len()));
+                // FIXME Using visual indenting, should use block or visual to match
+                // struct lit preference (however, in practice I think it is rare
+                // for struct patterns to be multi-line).
+                let offset = offset + path.len() + 3;
```
Comment (marcusklaas) [inline src/patterns.rs]: Could you add a comment explaining the origin of the 3 and 5?

File: src/patterns.rs [hunk 4]
```diff
@@ -55,39 +65,105 @@ impl Rewrite for Pat {
...
+                        field_string.push_str("..");
+                    } else {
+                        if field_string.len() > 0 {
+                            field_string.push_str(", ");
+                        }
+                        field_string.push_str("..");
+                    }
+                }
+
+                if field_string.len() == 0 {
```
Comment (marcusklaas) [inline src/patterns.rs]: Turbo nit: clippy will recommend `is_empty()`

File: src/patterns.rs [hunk 5]
```diff
@@ -97,3 +173,14 @@ impl Rewrite for Pat {
...
     }
 }
+
+impl Rewrite for FieldPat {
+    fn rewrite(&self, context: &RewriteContext, width: usize, offset: Indent) -> Option<String> {
+        let pat = self.pat.rewrite(context, width, offset);
+        if self.is_shorthand {
+            pat
+        } else {
+            Some(format!("{}: {}", self.ident.to_string(), try_opt!(pat)))
```
Comment (marcusklaas) [inline src/patterns.rs]: I'd wrap this in a `wrap_str` to be sure, or do a separate rewrite of the pat with a budget of `width - ident.to_string().len() - 2`.

File: src/patterns.rs [hunk 6]
```diff
@@ -15,20 +15,30 @@ use lists::{format_item_list, itemize_list};
...
                     BindingMode::BindByRef(mutability) => ("ref ", mutability),
                     BindingMode::BindByValue(mutability) => ("", mutability),
                 };
                 let mut_infix = format_mutability(mutability);
-                let result = format!("{}{}{}", prefix, mut_infix, ident.node);
+                let id_str = ident.node.to_string();
+
+                let sub_pat = match *sub_pat {
+                    Some(ref p) => {
+                        let width = try_opt!(width.checked_sub(prefix.len() + mut_infix.len() +
```
Comment (marcusklaas) [inline src/patterns.rs]: No `3` for `" @ "`?

File: src/patterns.rs (hunk 1 above)
Comment (nrc) [inline src/patterns.rs]: I always forget about this trick, but it makes the code so much nicer.

Comment (marcusklaas) [inline src/patterns.rs]: Right? I'm a big fan as well!

**Comment by cmbrandenburg** [comment]
Not as recent as the fix! Sorry for the dupe.

**Pull request merged**